apps/prairielearn/src/pages/instructorInstanceAdminSettings/instructorInstanceAdminSettings.js

```diff
@@ -17,136 +15,124 @@ import {
+      sql.short_names,
+      { course_id: res.locals.course.id },
+      z.array(z.string()),
+    );
+    res.locals.short_names = short_names;
 
-router.post('/', function (req, res, next) {
-  debug('POST /');
-  if (req.body.__action === 'copy_course_instance') {
-    debug('Copy course instance');
-    const editor = new CourseInstanceCopyEditor({
-      locals: res.locals,
-    });
-    editor.canEdit((err) => {
-      if (ERR(err, next)) return;
-      editor.doEdit((err, job_sequence_id) => {
-        if (ERR(err, (e) => logger.error('Error in doEdit()', e))) {
-          res.redirect(res.locals.urlPrefix + '/edit_error/' + job_sequence_id);
-        } else {
-          debug(
-            `Get course_instance_id from uuid=${editor.uuid} with course_id=${res.locals.course.id}`,
-          );
-          sqldb.queryOneRow(
-            sql.select_course_instance_id_from_uuid,
-            { uuid: editor.uuid, course_id: res.locals.course.id },
-            (err, result) => {
-              if (ERR(err, next)) return;
-              flash(
-                'success',
-                'Course instance copied successfully. You are new viewing your copy of the course instance.',
-              );
-              res.redirect(
-                res.locals.plainUrlPrefix +
-                  '/course_instance/' +
-                  result.rows[0].course_instance_id +
-                  '/instructor/instance_admin/settings',
-              );
-            },
-          );
-        }
+    const host = getCanonicalHost(req);
+    res.locals.studentLink = new URL(
+      res.locals.plainUrlPrefix + '/course_instance/' + res.locals.course_instance.id,
+      host,
+    ).href;
+
+    res.locals.studentLinkQRCode = new QR({
+      content: res.locals.studentLink,
+      width: 512,
+      height: 512,
+    }).svg();
+
+    res.locals.infoCourseInstancePath = encodePath(
+      path.join(
+        'courseInstances',
+        res.locals.course_instance.short_name,
+        'infoCourseInstance.json',
+      ),
+    );
+    res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+  }),
+);
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    if (req.body.__action === 'copy_course_instance') {
+      const editor = new CourseInstanceCopyEditor({
+        locals: res.locals,
       });
-    });
-  } else if (req.body.__action === 'delete_course_instance') {
-    debug('Delete course instance');
-    const editor = new CourseInstanceDeleteEditor({
-      locals: res.locals,
-    });
-    editor.canEdit((err) => {
-      if (ERR(err, next)) return;
-      editor.doEdit((err, job_sequence_id) => {
-        if (ERR(err, (e) => logger.error('Error in doEdit()', e))) {
-          res.redirect(res.locals.urlPrefix + '/edit_error/' + job_sequence_id);
-        } else {
-          res.redirect(`${res.locals.plainUrlPrefix}/course/${res.locals.course.id}/course_admin`);
+
+      const serverJob = await editor.prepareServerJob();
+      try {
```

Comment:
I'd consider structuring this a bit differently. As written, the `catch` block will also catch and silently swallow errors with `sqldb.queryRow` here. Really, we just want to wrap `executeWithServerJob` in `try`/`catch`. That is, something like this:

```ts
try {
  await editor.executeWithServerJob(serverJob);
} catch (err) {
  res.redirect(res.locals.urlPrefix + '/edit_error/' + serverJob.jobSequenceId);
  return;
}

const courseInstanceID = await sqldb.queryRow(...);
if (courseInstanceID === null) {
  throw new Error(...);
}

flash(...);
res.redirect(...);
```

Reply:
That makes sense. Updated the `try/catch` block to similarly reflect the above code in [7733056](https://github.com/PrairieLearn/PrairieLearn/pull/9591/commits/77330561c88ad62551104cdfe920201c8dc11abe).